kernel/src/fs/cgroupfs/controller/mod.rs

```diff
@@ -0,0 +1,413 @@
+}
+
+/// The type of sub-controller in the cgroup subsystem.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum ControllerType {
+    Memory,
+    CpuSet,
+    Pids,
+}
+
+impl ControllerType {
+    const ALL: [Self; 3] = [Self::Memory, Self::CpuSet, Self::Pids];
+}
+
+impl TryFrom<&str> for ControllerType {
+    type Error = aster_systree::Error;
+
+    fn try_from(value: &str) -> Result<Self> {
+        match value {
+            "memory" => Ok(ControllerType::Memory),
+            "cpuset" => Ok(ControllerType::CpuSet),
+            "pids" => Ok(ControllerType::Pids),
+            _ => Err(Error::NotFound),
+        }
+    }
+}
+
+bitflags! {
+    /// Bitflags representing active/deactive sub-control state.
+    pub(super) struct SubCtrlState: u8 {
+        const MEMORY_CTRLS = 1 << 0;
+        const CPUSET_CTRLS = 1 << 1;
+        const PIDS_CTRLS = 1 << 2;
+    }
+}
+
+impl SubCtrlState {
+    pub(super) fn control_bit(ctrl_type: ControllerType) -> Self {
+        match ctrl_type {
+            ControllerType::Memory => Self::MEMORY_CTRLS,
+            ControllerType::CpuSet => Self::CPUSET_CTRLS,
+            ControllerType::Pids => Self::PIDS_CTRLS,
+        }
+    }
+
+    /// Checks if a sub-control is active in the current state.
+    ///
+    /// If the given name does not represent a supported controller,
+    /// returns `None`.
+    pub(super) fn is_active(&self, ctrl_type: ControllerType) -> bool {
+        self.contains(Self::control_bit(ctrl_type))
+    }
+
+    fn activate(&mut self, ctrl_type: ControllerType) {
+        *self |= Self::control_bit(ctrl_type)
+    }
+
+    fn deactivate(&mut self, ctrl_type: ControllerType) {
+        *self -= Self::control_bit(ctrl_type)
+    }
+
+    pub(super) fn show(&self) -> String {
+        let mut controllers = Vec::new();
+
+        if self.contains(Self::MEMORY_CTRLS) {
+            controllers.push("memory");
+        }
+        if self.contains(Self::CPUSET_CTRLS) {
+            controllers.push("cpuset");
+        }
+        if self.contains(Self::PIDS_CTRLS) {
+            controllers.push("pids");
+        }
+
+        controllers.join(" ")
+    }
+
+    /// Returns an iterator over the active controller types.
+    pub(super) fn iter_active(&self) -> impl Iterator<Item = ControllerType> + '_ {
+        ControllerType::ALL
+            .into_iter()
+            .filter(|&ctrl_type| self.is_active(ctrl_type))
+    }
+}
+
+/// The sub-controller for a specific cgroup controller type.
+///
+/// If the sub-controller is inactive, the `inner` field will be `None`.
+struct SubController<T: SubControlStatic> {
+    inner: Option<T>,
+    /// The parent sub-controller in the hierarchy.
+    ///
+    /// This field is used to traverse the controller hierarchy.
+    #[expect(dead_code)]
+    parent: Option<Arc<SubController<T>>>,
+}
+
+impl<T: SubControl + SubControlStatic> SubController<T> {
```

Comment:
```suggestion
impl<T: SubControlStatic> SubController<T> {
```
`SubControl` is redundant.